force-app/main/default/classes/DMLOperations.cls

```diff
@@ -20,199 +20,345 @@
+    } else {
+      acct = new Account(Name = accountName, Description = 'New Account');
     }
-
-    /**
-     * Question 5
-     * Update the stage of an Opportunity in the database.
-     * Resource: https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql.htm
-     * @param oppId The Id of the Opportunity to update.
-     * @param newStage The new stage for the Opportunity.
-     */
-    public static void updateOpportunityStage(Id oppId, String newStage) {
-        // Retrieve the Opportunity from the database
-        // [SELECT Id FROM Opportunity WHERE Id = :oppId]
-
-        // Update the Opportunity's stage
-
-        // Update the Opportunity in the database
+    // Upsert the Account
+    upsert as user acct;
+    return acct;
+  }
+
+  /**
+   * Question 10
+   * Manage the association between Contacts and Accounts
+   * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
+   * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
+   * Example Contact.LastName = 'Doe' Link to Account.Name = 'Doe'
+   * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
+   * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
+   */
+  public static void upsertAccountsWithContacts(
+    List<Contact> contactsToUpsert
+  ) {
+    Set<String> acctNamesSet = new Set<String>();
+    Map<String, Id> acctNameToIdMap = new Map<String, Id>();
+    List<Account> acctsToInsert = new List<Account>();
+    // Loop through each Contact
+    for (Contact ctct : contactsToUpsert) {
+      // Extract the Account Name from Contact's LastName
+      acctNamesSet.add(ctct.LastName);
     }
-
-    /**
-     * Question 6
-     * Update the Account's Name and Industry given an Account Id.
-     * Resource: https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql.htm
-     * @param accountId The Id of the account to be updated.
-     * @param newName The new name for the account.
-     * @param newIndustry The new industry for the account.
-     */
-    public static void updateAccountFields(Id accountId, String newName, String newIndustry) {
-        // Query the account
-        // [SELECT Id FROM Account WHERE Id = :accountId]
-
-        // Update the Account with the given Id with the new name and industry
+    List<Account> existingAcctLst = [
+      SELECT Id, Name
+      FROM Account
+      WHERE Name IN :acctNamesSet
+    ];
+    String name;
+    List<String> acctNamesLst = new List<String>(acctNamesSet);
+    // Get the account based on the name or create one if it doesn't exist
+    for (Integer i = 0; i < acctNamesLst.size(); i++) {
```

Comment:
This is another scenario where using a Map can be a more powerful tool and enable you to avoid nested loops. You are still accomplishing what is needed and avoiding DML/SOQL inside the loops, which is great! Try utilizing a Map and see how this can be refactored to be more streamlined